SCMI server from SCP-firmware

etienne-lms opened the pull request:
The P-R allows a platform to build an SCMI server stack to run in OP-TEE core. The SCMI server is built from SCP-firmware reference  implementation https://github.com/ARM-software/SCP-firmware.git which source are to be fetch externally to optee_os and path set from `CFG_SCP_FIRMWARE`. 

This P-R implements an SCMI for platforms fvp and stm32mp this is embedded when `CFG_SCMI_SERVER=y`.
P-Rs https://github.com/OP-TEE/manifest/pull/220 and https://github.com/OP-TEE/build/pull/600 allows one to fetch and build stm32mp1 with SCP-firmware SCMI server.

jenswi-linaro commented on core/scmi-server/sub.mk:
```diff
@@ -0,0 +1,45 @@
+CFLAGS_OPTEE = $(cflags$(sm))
+CFLAGS_OPTEE += $(subst -include $(conf-file),,$(cflags$(sm)))
+CFLAGS_OPTEE += $(cflags-plat-scmi-server-y)
+scmi-server-flags-y += -DCFG_CFLAGS_OPTEE="$(CFLAGS_OPTEE)"
```
`$(cflags$(sm))` can change even after this `sub.mk` has been processed.
Have you verified that `$(CFLAGS_OPTEE)` is updated as needed if/when that happens?

etienne-lms replied:
Latest change do evaluate `$(cflagscore)` when target is executed, taking into account changes to its value after this _sub.mk_ file is parsed.

jenswi-linaro replied:
Thanks for checking.

jenswi-linaro commented on mk/config.mk:
```diff
@@ -742,6 +742,27 @@ $(eval $(call cfg-depends-all,CFG_SCMI_MSG_SMT_INTERRUPT_ENTRY,CFG_SCMI_MSG_SMT)
+endif #CFG_SCMI_SERVER
+
+ifeq ($(CFG_SCMI_MSG_DRIVERS)-$(CFG_SCMI_SERVER),y-y)
+$(error CFG_SCMI_MSG_DRIVERS=y and CFG_SCMI_SERVER=y are exclusive)
```
mutually exclusive

jenswi-linaro reviewed (commented):
Comments for "core: scmi_server: Build a SCMI server from SCP-firmware"
In the commit message:
`s/axternal/external/`

jenswi-linaro commented on core/arch/arm/plat-vexpress/conf.mk:
```diff
@@ -55,6 +55,11 @@ endif
 CFG_ENABLE_EMBEDDED_TESTS ?= y
 
+ifeq ($(CFG_SCMI_SERVER),y)
+CFG_SCMI_SERVER_PRODUCT=fvp
```
In the commit "plat-stm32mp1: support building with CFG_SCMI_SERVER=y" you're using force instead, any particular reason?

etienne-lms replied:
no, but here I should have used `$(call force,...)` or `?=`. I'll fix, using `$(call force,...)`.

jenswi-linaro commented:
For commit "drivers: clk: add support for clk_get_rates_array", please remove the `Change-Id:` tag.

xiaoxuZeng commented on core/scmi-server/scmi_server.c:
```diff
@@ -0,0 +1,86 @@
+	if (fwk_id < 0)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if (handle)
```
In the case `handle == NULL` I think it would be better to return `TEE_ERROR_BAD_PARAMETERS`. :)
Maybe:
```c
if (handle == NULL)
     return TEE_ERROR_BAD_PARAMETERS;

*handle = fwk_id;
```

xiaoxuZeng replied:
> Actaully there is a call to `scmi_server_get_channel()` from the SCMI PTA that only checks the channel ID is valid but does not retrieve the SCP-fmw related handle, [here in current patch](https://github.com/OP-TEE/optee_os/pull/5627/commits/8b96706fe7a5d7ed0033304fa3a5c5aec6997985#diff-b02cf9d13860a7b8eb818540e5776a98cba205eecc11519571d540cac8b5fa61R193-R196).

That‘s ok

etienne-lms commented on core/scmi-server/sub.mk:
```diff
@@ -0,0 +1,45 @@
+scmi-server-flags-y += -DCFG_CFLAGS_OPTEE="$(CFLAGS_OPTEE)"
+
+define build-SCP-firmware-cmake
+$(scp-firmware-output): FORCE
```
right. I think we can remove this.

etienne-lms replied:
We need `FORCE`. We need to run CMake on SCP-firmware to know if something needs to be rebuilt, optee_ios as no visibility on whether its source file tree content was modified in some ways.

jenswi-linaro replied:
Fair enough, I see that you at least got rid of the `rm -rf`

xiaoxuZeng reviewed (approved):
lgtm

etienne-lms commented:
@xiaoxuZeng, can I add your Acked-by or Reviewed-by tag to the commits?
To do so I would need your tag ID [`Full Name <mail-address>`](https://optee.readthedocs.io/en/latest/general/contribute.html#developer-certificate-of-origin).

xiaoxuZeng commented:
> @xiaoxuZeng, can I add your Acked-by or Reviewed-by tag to the commits? To do so I would need your tag ID [`Full Name <mail-address>`](https://optee.readthedocs.io/en/latest/general/contribute.html#developer-certificate-of-origin).
Sure

Reviewed-by: Xiaoxu Zeng <zengxiaoxu@huawei.com>

thanks.

etienne-lms commented:
Tag appllied + fix include order:
```patch
--- a/core/scmi-server/scmi_server.c
+++ b/core/scmi-server/scmi_server.c
@@ -8,5 +8,5 @@
 #include <initcall.h>
 #include <kernel/panic.h>
 #include <mm/core_memprot.h>
+#include <optee_scmi.h>
 #include <scmi/scmi_server.h>
-#include <optee_scmi.h>

etienne-lms commented:
rebased

etienne-lms commented:
I've updated _ci.yml_ but it requires CMake. Can you add it in your optee_os_ci image?
CI messsage:
```
if [ -d $HOME/scp-firmware ]; then _make PLATFORM=stm32mp1-157C_DK2 CFG_SCMI_SERVER=y CFG_SCP_FIRMWARE=$HOME/scp-firmware; fi
bash: line 1: cmake: command not found
make: *** [core/scmi-server/sub.mk:28: out/core/scmi-server/build/product/optee-stm32mp1/fw/libscmi-fw-all.a] Error 127
Error: Process completed with exit code 2.

jforissier commented:
> I've updated _ci.yml_ but it requires CMake. Can you add it in your optee_os_ci image?

Done. For commit "ci: build stm32mp15 with scp-firmware scmi-server":

`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

etienne-lms commented:
Rebased to trigger a new CI round.

etienne-lms commented:
CI build failed. Some missing glibc things:
```
fwk_id.c:(.text.__fwk_id_str+0x72): undefined reference to `__snprintf_chk'
/usr/bin/arm-linux-gnueabihf-ld.bfd: fwk_id.c:(.text.__fwk_id_str+0x12e): undefined reference to `__snprintf_chk'
fwk_io.c:(.text.fwk_io_vprintf+0x30): undefined reference to `__vsnprintf_chk'
```
Browsing a bit this could be related to use of musl in the doccker image.

jforissier commented:
> Browsing a bit this could be related to use of musl in the doccker image.

Does it build successfully for you? Can you give the full command line?
CI doesn't use the GCC compiler built by Arm (the one we get via `make toolchains` in build git), it might be the reason. That being said, which libc the compiler provides should not make a difference since OP-TEE has its own (libutils).

jforissier commented:
It looks like the C flags for the SCP firmware are missing `-nostdinc`. Therefore the compiler uses the headers for the default C library instead of the OP-TEE ones (from libutils). Refer to the objdump output of `fwk_log.c.obj`:
```
fwk_log_vsnprintf():
/tmp/SCP-firmware/framework/src/fwk_log.c:124
  46:   9701            str     r7, [sp, #4]
snprintf():
/usr/arm-linux-gnueabihf/include/bits/stdio2.h:71
  48:   f7ff fffe       bl      0 <__snprintf_chk>
                        48: R_ARM_THM_CALL      __snprintf_chk
  4c:   682a            ldr     r2, [r5, #0]
  4e:   4605            mov     r5, r0
```

The file `/usr/arm-linux-gnueabihf/include/bits/stdio2.h` should not be used. Instead ``snprintf()` should be found in `lib/libutils/isoc/include/stdio.h`. Since libutils provides no implementation for `__snprintf_chk()` there is a link error.

etienne-lms commented:
> It looks like the C flags for the SCP firmware are missing -nostdinc. (...)

Many thanks for the analysis and to the OP-TEE CI that issue was reported.

I've have changed _scmi-server/sub.mk_ to set core nostdinc and libutils include paths to CFLAGS directives passed through CMake.

Changing SCP-firmware to build with nostdinc, i faced 2 issues.
First _assert.h_ and _stdlib.h_ header files from SCP-fmw and optee_os are using the same `ASSERT_H` / `STDLIB_H` marco name for header guards. I've added an RFC commit to this P-R describing the issue and a change proposal in our header files.
Second is `aligned_alloc()` needed by SCP-firmware. As is it a standard function, the easiest way to address its declaration and implementation is to implement it in libutils. See related RFC commit.

I've updated _ci.yml_ in the series to point to my SCP-firmare branch to haved needed changes in that repo and check CI results. I'll sync with @vingu-linaro to update P-R https://github.com/ARM-software/SCP-firmware/pull/664 with related changes.

jforissier commented on lib/libutils/isoc/include/stdlib.h:
```diff
@@ -15,6 +15,8 @@
 #include <stddef.h>
 #include <malloc.h>
 
+/* aligend_alloc() is declared in included malloc.h */
```
I would prefer to keep things as close to the standard as possible, that is: declare `aligned_alloc()` in `<stdlib.h>` not in `<malloc.h>` and use the proper C11 guard:
```C
#if __STDC_VERSION__ >= 201112L
void *aligned_alloc(size_t alignment, size_t size);
#endif
```

[Upstream newlib](https://github.com/bminor/newlib/blob/master/newlib/libc/include/stdlib.h#L331) uses `__ISO_C_VISIBLE` which seems to be the BSD libc way of doing things, I think it makes more sense to use `__STDC_VERSION__ defined by the compiler.

jforissier reviewed (changes requested):
Reviewing "[RFC] libutils: implement aligned_alloc()"

jforissier commented:
Commit "[RFC] libutils: prefix system header guard for assert.h and stdlib.h":

LGTM, but it means all our libc (libutils) header should probably do the same... we can leave it until another conflict appears though.

`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

etienne-lms commented:
I can apply the change to all std header from libutils. I also think it would be more consistent.

jforissier commented:
> I can apply the change to all std header from libutils. I also think it would be more consistent.

Fine by me.

etienne-lms commented:
Is `#if __STDC_VERSION__ >= 201112L` really needed to protect `aligned_alloc()` declaration? We can implement to function whatever the toolchain used, no?
Moreover I saw libutils build for ldelf fails when we have these guards.

etienne-lms commented:
I have added fixup commits to move `aligned_alloc()` declaration to _stdlib.h_ (without guards)
and to add prefix `_` (single underscore char) to standard header files (not optee specific _trace.h_, _util.h_, etc...

jforissier commented:
> Is `#if __STDC_VERSION__ >= 201112L` really needed to protect `aligned_alloc()` declaration? We can implement to function whatever the toolchain used, no? Moreover I saw libutils build for ldelf fails when we have these guards.

It is a dangerous path to take IMO. What if some external code depends on this (which is a standard) to provide its own implementation? Then we would have a conflict.

etienne-lms commented:
We build with `-std=gnu99`, not c11. Does that mean we can't implement `aligned_alloc()` in libutils?

jforissier commented:
I suppose we can easily switch to `-std=c11` or `std=gnu11`.

etienne-lms commented:
I've sqaushed the fixup commits and rebased the series since `aligned_alloc()` is merged.
I've added 3 new fixup commits related to discussion with @vingu-linaro.

@xiaoxuZeng and @jforissier, can you check these 3 added changes. I'll apply your review tag upon your feedback.

jforissier commented on core/scmi-server/sub.mk:
```diff
@@ -14,6 +14,7 @@ incdirs_ext-y += $(scmi-server-scp-path)/arch/none/optee/include
 scp-firmware-output = $(scmi-server-out-path)/build/product/optee-$(scmi-server-product)/fw/libscmi-fw-all.a
 
 libdeps += $(scp-firmware-output)
+libdeps += out/arm/core-lib/libutils/libutils.a
```
You don't want to hard-code the output directory here! It is the reason of the CI failure, which uses a custom output directory ([link](https://github.com/OP-TEE/optee_os/actions/runs/3542522124/jobs/5948110592#step:5:183)).

The fix should be simple enough though. I suggest to define `lib-libfile-$(libname)-$(sm)` similarly to the `-shlibfile-` version which already exists:
```diff
diff --git a/mk/lib.mk b/mk/lib.mk
index 30f2040a3..65b575ec5 100644
--- a/mk/lib.mk
+++ b/mk/lib.mk
@@ -27,6 +27,7 @@ endif
 include mk/compile.mk

 lib-libfile    := $(out-dir)/$(base-prefix)$(libdir)/lib$(libname).a
+lib-libfile-$(libname)-$(sm) := $(lib-libfile)
 ifeq ($(CFG_ULIBS_SHARED),y)
 lib-shlibfile  := $(out-dir)/$(base-prefix)$(libdir)/lib$(libname).so
 lib-shlibstrippedfile := $(out-dir)/$(base-prefix)$(libdir)/lib$(libname).stripped.so
```

Then you should be able to use: `libdeps += $(lib-libfile-libutils-core)` here.

etienne-lms replied:
> EDIT: But why is this dependency even needed? libscmi-fw-all.a is an archive, it should not need to be linked against libutils.

Actually... you're right, we don't need it. Libutils is already in the system deps, as the scim-server lib we're building. I'll crosscheck with @vingu-linaro, maybe related to a specific build env.

vingu-linaro replied:
It's not about linking with archive but the order of the lib when optee_os is linked into the final binary.
Without this dependency, libutils.a is before libscmi-fw-all.a some functions of libutils are removed because not used at this point. With this, we ensure that libscmi-fw-all.a is before and add a ref to libutils.a function

jenswi-linaro replied:
> It's not about linking with archive but the order of the lib when optee_os is linked into the final binary. Without this dependency, libutils.a is before libscmi-fw-all.a some functions of libutils are removed because not used at this point. With this, we ensure that libscmi-fw-all.a is before and add a ref to libutils.a function

That is surprising, `libutils.a` is usually the last lib (expect `libgcc.a`) on the command line.

jforissier replied:
> It's not about linking with archive but the order of the lib when optee_os is linked into the final binary.

Right, I was a bit confused by the name (`libdeps`), as if the library was needed to produce the SCP .a. It's not -- very well.

> That is surprising, `libutils.a` is usually the last lib (except `libgcc.a`) on the command line.

Correct, this is supposed to be achieved by:
```
mk/lib.mk:libdeps               := $(lib-libfile) $(libdeps)
```
...and the fact that `libutils` is listed first in `core.mk`:

```
# Do libraries
...
libname = utils
libdir = lib/libutils
include mk/lib.mk 
...
ifeq ($(CFG_SCMI_SERVER),y)
libname = scmi-server
libdir = core/scmi-server
include mk/lib.mk
endif
```

vingu-linaro replied:
without this line i have : 
/usr/bin/ccache /home/vingu/Linaro/SCMI/optee-project-qemu-latest/build/../toolchains/aarch64/bin/aarch64-linux-gnu-ld.bfd -T core/arch/arm/kernel/link_dummy.ld --no-check-sections  out/arm/core/arch/arm/kernel/ldelf_loader.o out/arm/core/arch/arm/kernel/tee_time.o out/arm/core/arch/arm/kernel/rpc_io_i2c.o out/arm/core/arch/arm/kernel/otp_stubs.o out/arm/core/arch/arm/kernel/delay.o out/arm/core/arch/arm/kernel/idle.o out/arm/core/arch/arm/kernel/tee_time_arm_cntpct.o out/arm/core/arch/arm/kernel/timer_a64.o out/arm/core/arch/arm/kernel/spin_lock_a64.o out/arm/core/arch/arm/kernel/tlb_helpers_a64.o out/arm/core/arch/arm/kernel/cache_helpers_a64.o out/arm/core/arch/arm/kernel/thread_a64.o out/arm/core/arch/arm/kernel/thread.o out/arm/core/arch/arm/kernel/thread_optee_smc.o out/arm/core/arch/arm/kernel/thread_optee_smc_a64.o out/arm/core/arch/arm/kernel/abort.o out/arm/core/arch/arm/kernel/vfp.o out/arm/core/arch/arm/kernel/vfp_a64.o out/arm/core/arch/arm/kernel/trace_ext.o out/arm/core/arch/arm/kernel/misc_a64.o out/arm/core/arch/arm/kernel/boot.o out/arm/core/arch/arm/kernel/entry_a64.o out/arm/core/arch/arm/kernel/unwind_arm64.o out/arm/core/arch/arm/mm/core_mmu.o out/arm/core/arch/arm/mm/core_mmu_lpae.o out/arm/core/arch/arm/mm/pgt_cache.o out/arm/core/arch/arm/mm/mobj_dyn_shm.o out/arm/core/arch/arm/tee/arch_svc_a64.o out/arm/core/arch/arm/tee/arch_svc.o out/arm/core/arch/arm/tee/entry_fast.o out/arm/core/arch/arm/tee/cache.o out/arm/core/arch/arm/plat-vexpress/main.o out/arm/core/arch/arm/plat-vexpress/vendor_props.o out/arm/core/crypto/crypto.o out/arm/core/crypto/aes-gcm.o out/arm/core/crypto/aes-gcm-sw.o out/arm/core/crypto/aes-gcm-ghash-tbl.o out/arm/core/crypto/signed_hdr.o out/arm/core/crypto/rng_fortuna.o out/arm/core/crypto/cbc-mac.o out/arm/core/crypto/aes-cts.o out/arm/core/crypto/sm2-kdf.o out/arm/core/crypto/sm3.o out/arm/core/crypto/sm3-hash.o out/arm/core/crypto/sm3-hmac.o out/arm/core/crypto/sm4.o out/arm/core/crypto/sm4-ecb.o out/arm/core/crypto/sm4-cbc.o out/arm/core/crypto/sm4-ctr.o out/arm/core/drivers/pl011.o out/arm/core/drivers/gic.o out/arm/core/drivers/clk/clk.o out/arm/core/drivers/clk/clk_dt.o out/arm/core/drivers/clk/fixed_clk.o out/arm/core/drivers/rstctrl/rstctrl.o out/arm/core/kernel/spin_lock_debug.o out/arm/core/kernel/assert.o out/arm/core/kernel/console.o out/arm/core/kernel/dt.o out/arm/core/kernel/dt_driver.o out/arm/core/kernel/dt_driver_test.o out/arm/core/kernel/pm.o out/arm/core/kernel/handle.o out/arm/core/kernel/interrupt.o out/arm/core/kernel/ldelf_syscalls.o out/arm/core/kernel/msg_param.o out/arm/core/kernel/panic.o out/arm/core/kernel/refcount.o out/arm/core/kernel/tee_misc.o out/arm/core/kernel/tee_ta_manager.o out/arm/core/kernel/ts_manager.o out/arm/core/kernel/scattered_array.o out/arm/core/kernel/huk_subkey.o out/arm/core/kernel/user_mode_ctx.o out/arm/core/kernel/initcall.o out/arm/core/kernel/user_access.o out/arm/core/kernel/mutex.o out/arm/core/kernel/wait_queue.o out/arm/core/kernel/notif.o out/arm/core/kernel/thread.o out/arm/core/kernel/user_ta.o out/arm/core/kernel/ree_fs_ta.o out/arm/core/kernel/early_ta.o out/arm/core/kernel/secstor_ta.o out/arm/core/kernel/embedded_ts.o out/arm/core/kernel/pseudo_ta.o out/arm/core/mm/mobj.o out/arm/core/mm/fobj.o out/arm/core/mm/file.o out/arm/core/mm/vm.o out/arm/core/mm/core_mmu.o out/arm/core/mm/tee_mm.o out/arm/core/pta/device.o out/arm/core/pta/secstor_ta_mgmt.o out/arm/core/pta/stats.o out/arm/core/pta/system.o out/arm/core/pta/scmi.o out/arm/core/pta/tests/fs_htree.o out/arm/core/pta/tests/invoke.o out/arm/core/pta/tests/misc.o out/arm/core/pta/tests/mutex.o out/arm/core/pta/tests/aes_perf.o out/arm/core/tee/entry_std.o out/arm/core/tee/tee_cryp_utl.o out/arm/core/tee/tee_cryp_hkdf.o out/arm/core/tee/tee_cryp_concat_kdf.o out/arm/core/tee/tee_cryp_pbkdf2.o out/arm/core/tee/tee_obj.o out/arm/core/tee/tee_svc.o out/arm/core/tee/tee_svc_cryp.o out/arm/core/tee/tee_svc_storage.o out/arm/core/tee/tee_time_generic.o out/arm/core/tee/tadb.o out/arm/core/tee/socket.o out/arm/core/tee/tee_ta_enc_manager.o out/arm/core/tee/tee_fs_key_manager.o out/arm/core/tee/tee_ree_fs.o out/arm/core/tee/fs_dirfile.o out/arm/core/tee/fs_htree.o out/arm/core/tee/tee_fs_rpc.o out/arm/core/tee/tee_pobj.o out/arm/core/tee/uuid.o out/arm/core/tee/tee_supp_plugin_rpc.o out/arm/core/ta_pub_key.o out/arm/core/ldelf_hex.o out/arm/core/early_ta_f04a0fe7-1f5d-4b9b-abf7-619b85b4ce8c.o out/arm/core/arch/arm/dts/embedded_dtb_test.o   out/arm/core/scmi-server/libscmi-server.a out/arm/lib/libunw/libunw.a out/arm/core/lib/zlib/libzlib.a out/arm/core/lib/libfdt/libfdt.a out/arm/core/lib/libtomcrypt/libtomcrypt.a out/arm/core-lib/libmbedtls/libmbedtls.a out/arm/core-lib/libutils/libutils.a  out/arm/core/scmi-server/build/product/optee-fvp/fw/libscmi-fw-all.a /home/vingu/Linaro/SCMI/optee-project-qemu-latest/toolchains/aarch64/bin/../lib/gcc/aarch64-none-linux-gnu/10.2.1/libgcc.a -o out/arm/core/all_objs.o
/home/vingu/Linaro/SCMI/optee-project-qemu-latest/build/../toolchains/aarch64/bin/aarch64-linux-gnu-ld.bfd: /home/vingu/Linaro/SCMI/optee-project-qemu-latest/optee_os/out/arm/core/scmi-server/build/framework/libframework.a(fwk_string.c.obj): in function `fwk_str_strncpy':
/home/vingu/Linaro/SCMI/optee-project-qemu-latest/SCP-firmware/framework/src/fwk_string.c:52: undefined reference to `strncpy'
core/arch/arm/kernel/link.mk:68: recipe for target 'out/arm/core/all_objs.o' failed

With the line it becomes: 

/usr/bin/ccache /home/vingu/Linaro/SCMI/optee-project-qemu-latest/build/../toolchains/aarch64/bin/aarch64-linux-gnu-ld.bfd -T core/arch/arm/kernel/link_dummy.ld --no-check-sections  out/arm/core/arch/arm/kernel/ldelf_loader.o out/arm/core/arch/arm/kernel/tee_time.o out/arm/core/arch/arm/kernel/rpc_io_i2c.o out/arm/core/arch/arm/kernel/otp_stubs.o out/arm/core/arch/arm/kernel/delay.o out/arm/core/arch/arm/kernel/idle.o out/arm/core/arch/arm/kernel/tee_time_arm_cntpct.o out/arm/core/arch/arm/kernel/timer_a64.o out/arm/core/arch/arm/kernel/spin_lock_a64.o out/arm/core/arch/arm/kernel/tlb_helpers_a64.o out/arm/core/arch/arm/kernel/cache_helpers_a64.o out/arm/core/arch/arm/kernel/thread_a64.o out/arm/core/arch/arm/kernel/thread.o out/arm/core/arch/arm/kernel/thread_optee_smc.o out/arm/core/arch/arm/kernel/thread_optee_smc_a64.o out/arm/core/arch/arm/kernel/abort.o out/arm/core/arch/arm/kernel/vfp.o out/arm/core/arch/arm/kernel/vfp_a64.o out/arm/core/arch/arm/kernel/trace_ext.o out/arm/core/arch/arm/kernel/misc_a64.o out/arm/core/arch/arm/kernel/boot.o out/arm/core/arch/arm/kernel/entry_a64.o out/arm/core/arch/arm/kernel/unwind_arm64.o out/arm/core/arch/arm/mm/core_mmu.o out/arm/core/arch/arm/mm/core_mmu_lpae.o out/arm/core/arch/arm/mm/pgt_cache.o out/arm/core/arch/arm/mm/mobj_dyn_shm.o out/arm/core/arch/arm/tee/arch_svc_a64.o out/arm/core/arch/arm/tee/arch_svc.o out/arm/core/arch/arm/tee/entry_fast.o out/arm/core/arch/arm/tee/cache.o out/arm/core/arch/arm/plat-vexpress/main.o out/arm/core/arch/arm/plat-vexpress/vendor_props.o out/arm/core/crypto/crypto.o out/arm/core/crypto/aes-gcm.o out/arm/core/crypto/aes-gcm-sw.o out/arm/core/crypto/aes-gcm-ghash-tbl.o out/arm/core/crypto/signed_hdr.o out/arm/core/crypto/rng_fortuna.o out/arm/core/crypto/cbc-mac.o out/arm/core/crypto/aes-cts.o out/arm/core/crypto/sm2-kdf.o out/arm/core/crypto/sm3.o out/arm/core/crypto/sm3-hash.o out/arm/core/crypto/sm3-hmac.o out/arm/core/crypto/sm4.o out/arm/core/crypto/sm4-ecb.o out/arm/core/crypto/sm4-cbc.o out/arm/core/crypto/sm4-ctr.o out/arm/core/drivers/pl011.o out/arm/core/drivers/gic.o out/arm/core/drivers/clk/clk.o out/arm/core/drivers/clk/clk_dt.o out/arm/core/drivers/clk/fixed_clk.o out/arm/core/drivers/rstctrl/rstctrl.o out/arm/core/kernel/spin_lock_debug.o out/arm/core/kernel/assert.o out/arm/core/kernel/console.o out/arm/core/kernel/dt.o out/arm/core/kernel/dt_driver.o out/arm/core/kernel/dt_driver_test.o out/arm/core/kernel/pm.o out/arm/core/kernel/handle.o out/arm/core/kernel/interrupt.o out/arm/core/kernel/ldelf_syscalls.o out/arm/core/kernel/msg_param.o out/arm/core/kernel/panic.o out/arm/core/kernel/refcount.o out/arm/core/kernel/tee_misc.o out/arm/core/kernel/tee_ta_manager.o out/arm/core/kernel/ts_manager.o out/arm/core/kernel/scattered_array.o out/arm/core/kernel/huk_subkey.o out/arm/core/kernel/user_mode_ctx.o out/arm/core/kernel/initcall.o out/arm/core/kernel/user_access.o out/arm/core/kernel/mutex.o out/arm/core/kernel/wait_queue.o out/arm/core/kernel/notif.o out/arm/core/kernel/thread.o out/arm/core/kernel/user_ta.o out/arm/core/kernel/ree_fs_ta.o out/arm/core/kernel/early_ta.o out/arm/core/kernel/secstor_ta.o out/arm/core/kernel/embedded_ts.o out/arm/core/kernel/pseudo_ta.o out/arm/core/mm/mobj.o out/arm/core/mm/fobj.o out/arm/core/mm/file.o out/arm/core/mm/vm.o out/arm/core/mm/core_mmu.o out/arm/core/mm/tee_mm.o out/arm/core/pta/device.o out/arm/core/pta/secstor_ta_mgmt.o out/arm/core/pta/stats.o out/arm/core/pta/system.o out/arm/core/pta/scmi.o out/arm/core/pta/tests/fs_htree.o out/arm/core/pta/tests/invoke.o out/arm/core/pta/tests/misc.o out/arm/core/pta/tests/mutex.o out/arm/core/pta/tests/aes_perf.o out/arm/core/tee/entry_std.o out/arm/core/tee/tee_cryp_utl.o out/arm/core/tee/tee_cryp_hkdf.o out/arm/core/tee/tee_cryp_concat_kdf.o out/arm/core/tee/tee_cryp_pbkdf2.o out/arm/core/tee/tee_obj.o out/arm/core/tee/tee_svc.o out/arm/core/tee/tee_svc_cryp.o out/arm/core/tee/tee_svc_storage.o out/arm/core/tee/tee_time_generic.o out/arm/core/tee/tadb.o out/arm/core/tee/socket.o out/arm/core/tee/tee_ta_enc_manager.o out/arm/core/tee/tee_fs_key_manager.o out/arm/core/tee/tee_ree_fs.o out/arm/core/tee/fs_dirfile.o out/arm/core/tee/fs_htree.o out/arm/core/tee/tee_fs_rpc.o out/arm/core/tee/tee_pobj.o out/arm/core/tee/uuid.o out/arm/core/tee/tee_supp_plugin_rpc.o out/arm/core/ta_pub_key.o out/arm/core/ldelf_hex.o out/arm/core/early_ta_f04a0fe7-1f5d-4b9b-abf7-619b85b4ce8c.o out/arm/core/arch/arm/dts/embedded_dtb_test.o   out/arm/core/scmi-server/libscmi-server.a out/arm/lib/libunw/libunw.a out/arm/core/lib/zlib/libzlib.a out/arm/core/lib/libfdt/libfdt.a out/arm/core/lib/libtomcrypt/libtomcrypt.a out/arm/core-lib/libmbedtls/libmbedtls.a out/arm/core-lib/libutils/libutils.a  out/arm/core/scmi-server/build/product/optee-fvp/fw/libscmi-fw-all.a out/arm/core-lib/libutils/libutils.a /home/vingu/Linaro/SCMI/optee-project-qemu-latest/toolchains/aarch64/bin/../lib/gcc/aarch64-none-linux-gnu/10.2.1/libgcc.a -o out/arm/core/all_objs.o

The only difference is the position of libscmi-fw-all.a vs libutils.a

etienne-lms replied:
moving `scmi-server` before `libutils` in _core/core.mk_ would resolve the dependency. But maybe a bit fragile.

vingu-linaro replied:
I have also considered this change but I agree that it's a bit fragile

jforissier replied:
> moving `scmi-server` before `libutils` in _core/core.mk_ would resolve the dependency. But maybe a bit fragile.

The libs should be in reverse order of appearance in core.mk. `mk/lib.mk` should take care of that, perhaps there is a bug in it or in the way it is used.

The naming is a bit weird too, I see no reason not stick more closely with the lib model and have all this code under `core/lib/libscmi-fwk-all`. Any deviation can only be painful when looking for stuff etc.

jenswi-linaro replied:
The problem is one line above. `mk/lib.mk` adds to `libdeps` in front and here you're appending.

etienne-lms replied:
thanks @jenswi-linaro, indeed changing line 16 should fix your issue @vingu-linaro:
```patch
-libdeps += $(scp-firmware-output)
+libdeps := $(scp-firmware-output) $(libdeps)
```

I'll push a fixup commit.

jforissier reviewed (commented):
@etienne-lms one comment, once addressed LGTM.

etienne-lms commented:
@vingu-linaro, I've added your change to add support for qemuv8.

etienne-lms commented:
I've pushed 2 more fixup commits: moving _scmi-server/_ to _core/lib/_ and fixing the issue discussed in https://github.com/OP-TEE/optee_os/pull/5627#discussion_r1031737099. 
@vingu-linaro: can you update https://github.com/ARM-software/SCP-firmware/pull/664 with this change
https://github.com/etienne-lms/SCP-firmware/commit/454eb02250a92f5cc5299d2b794e4f88368745f2?

CI should succeed to build.

jforissier commented:
> CI should succeed to build.

It does! :+1: 

Honestly this is somewhat abusing the `mk/lib.mk` mechanism, because the build rules are not supposed to be in the `sub.mk` file (i.e., `core/lib/scmi-server/sub.mk`). As a result `mk/lib.mk` does things that are not needed. A cleaner approach would be to create a dedicated file for external sources built using cmake, such as `mk/lib_ext_cmake.mk`, which could be fed a couple of parameters form the `sub.mk`. But I'm OK with this until we meet another such case.

jenswi-linaro commented:
> A cleaner approach would be to create a dedicated file for external sources built using cmake, such as `mk/lib_ext_cmake.mk`, which could be fed a couple of parameters form the `sub.mk`. But I'm OK with this until we meet another such case.

I disagree, a clean approach would include the needed parts of SCP-firmware in the optee_os git and compile it as any other library. The way the se050 stuff is done is a compromise, but at least we have a well-integrated build system.

This interaction with cmake is terrible, it's bad enough to build with a sub-make, and the only way to make it worse would be to include autotools in the stew.

jforissier commented:
> > A cleaner approach would be to create a dedicated file for external sources built using cmake, such as `mk/lib_ext_cmake.mk`, which could be fed a couple of parameters form the `sub.mk`. But I'm OK with this until we meet another such case.
> 
> I disagree, a clean approach would include the needed parts of SCP-firmware in the optee_os git and compile it as any other library. The way the se050 stuff is done is a compromise, but at least we have a well-integrated build system.

Hmmm. Yeah. But it seems a bit much to ask the SCP Firmware project to adopt the OP-TEE build system, because OP-TEE is only one possible target, right?

> This interaction with cmake is terrible, it's bad enough to build with a sub-make, and the only way to make it worse would be to include autotools in the stew.

Hey, bitbake would be even worse! :wink:

jenswi-linaro commented:
> Hmmm. Yeah. But it seems a bit much to ask the SCP Firmware project to adopt the OP-TEE build system, because OP-TEE is only one possible target, right?

That hasn't stopped us with mbedtls, LTC, zlib, libfdt and se050.

etienne-lms commented:
Fixup commit squashed.
I applied @jforissier review tag.

@xiaoxuZeng, could you have a look at the 2 below commits?
_"core: scmi_server: Build a SCMI server from SCP-firmware"_
_"core: pta: scmi: support SCP-firmware SCMI resources"_
These commits have changed since your review tag was sent so I removed your tag from these 2 commits. Please, tell me if I can apply you review tag.

etienne-lms commented:
After dicsussions with maintainers, it's been decided to change how we build SCP-firmware, not using CMake native build of the package but rather using optee_os sub.mk based makefiles. I've created https://github.com/OP-TEE/optee_os/pull/5686 for that purpose and will close current P-R.
